src/utils/basic-auth.ts

```diff
@@ -0,0 +1,27 @@
+import type { HonoRequest } from '../request'
+import { decodeBase64 } from './encode'
+
+const CREDENTIALS_REGEXP = /^ *(?:[Bb][Aa][Ss][Ii][Cc]) +([A-Za-z0-9._~+/-]+=*) *$/
+const USER_PASS_REGEXP = /^([^:]*):(.*)$/
+const utf8Decoder = new TextDecoder()
+
+export type Auth = (req: HonoRequest) => { username: string; password: string } | undefined
```

Comment:
I prefer making `req` as `Request` not `HonoRequest`. The utility function should be more general. So, `HonoRequest` that depends on Hono is not good.

Reply:
Thank you for the review.
I have removed the dependency on HonoRequest.
https://github.com/honojs/hono/pull/3359/commits/989a0db600e2a3942b1e2257463b6fa7669a714f